etc/migration/migrateUsersToShibboleth.js

```diff
@@ -0,0 +1,388 @@
+
+
+// Spin up the application container. This will allow us to re-use existing APIs
+OAE.init(config, function(err) {
+    if (err) {
+        log().error({'err': err}, 'Unable to spin up the application server');
+        return _exit(err.code);
+    }
+
+    _getAllUsersForTenant(tenantAlias, function(err, users) {
+        _mapUsersToShibUsers(users, function(err, mappedUsers) {
+            if (err) {
+                log().error({'err':err}, 'Encountered error when migrating %s users to Shibboleth', tenantAlias);
+                return _exit(1);
+            }
+
+            if (errors > 0) {
+                log().warn('Some users were not mapped to Shibboleth users, check the CSV file for more information');
+            }
+
+            log().info('%s users were migrated to Shibboleth login.', mappedUsers.length);
+            return _exit(0);
+        });
+    });
+});
+
+/**
+ * Page through all the users in the system, and filter them by tenant
+ *
+ * @param  {String}     tenantAlias             The tenant alias to filter by
+ * @param  {Function}   callback                Invoked when users have been collected
+ * @param  {Object}     callback.err            The error that occurred, if any
+ * @param  {Object}     callback.userHashes     An array of user hashes
+ * @api private
+ */
+var _getAllUsersForTenant = function(tenantAlias, callback) {
+    var userHashes = [];
+
+    // Just get the fields we'll need later
+    PrincipalsDAO.iterateAll(['principalId', 'tenantAlias', 'email'], 100, _aggregateUsers, function(err) {
+        if (err) {
+            log().error({'err': err}, 'Failed to iterate all users for tenant %s', tenantAlias);
+            return _exit(1);
+        }
+
+        log().info('Found %s users for specified tenant', userHashes.length);
+
+        return callback(null, userHashes);
+    });
+
+    /*
+     * Filter users down to those that are part of the specified tenant. Then add them to the `userHashes` array
+     *
+     * @param  {Object[]}   principalHashes     The principals to filter and aggregate
+     * @param  {Function}   callback            Will be invoked when the principals are aggregated
+     */
+    function _aggregateUsers(principalHashes, callback) {
+        log().info('Checking %s principals for tenancy', principalHashes.length);
+
+        _.chain(principalHashes)
+            .filter(function(principalHash) {
+                return principalHash.tenantAlias === tenantAlias && PrincipalsDAO.isUser(principalHash.principalId) && principalHash.email;
+            })
+            .each(function(userHash) {
+                // Add the external ID for each user
+                _findExternalId(userHash, function(err, userHash) {
+                    if (err) {
+                        return callback(err);
+                    }
+
+                    log().info('Adding user %s to tenant users', userHash.email);
+                    userHashes.push(userHash);
+                });
```

Comment:
This here is a bit of an anti-pattern, and can cause unreliable results with respect to asynchronous processing and race conditions.

Essentially what you're doing is, for each principal in the `principalHashes`, executing a query that does not wait for the return. This can potentially result in thousands of in-flight requests waiting for response from the database. Notice how the `.each` doesn't have a callback to indicate when to continue processing, that's kind of what tips off what's happening here.

What you might want to try instead is a recursive function that starts with the array of principals hashes, and executes `_findExternalId` for each principal, then invoking the next recursive step after the external id has been accessed for each principal. Here's a rough example of looping over asynchronous callbacks:

```javascript
function _aggregateUsers(principalHashes, callback, _userHashes) {
  _userHashes = _userHashes || []
  if (_.isEmpty(principalHashes)) {
    // This is our base case, when there are no principal hashes left
    return callback(null, principalHashes);
  }

  // Take the next principal hash from the collection
  var principalHash = principalHashes.shift();
  _findExternalId(principalHash, function(err, userHash) {
    if (err) {
      return callback(err);
    }

    // Accumulate the return information we want
    _userHashes.push(userHash);

    // Our recursive step inside the callback for _findExternalId
    return _aggregateUsers(principalHashes, callback, userHashes)
  });
}
```

There are tools like `async` that makes these patterns a bit easier as well, however we don't really use it yet.

This pattern could cause some of the strange results you've been seeing.